### ISSUE TITLE
Issues with Unreal Engine generated projects

### ISSUE THREAD

**Issue by romanholidaypancakes** [microsoft/vscode-cpptools#8812]
Bug type: Language Service

**Describe the bug**
![image](https://user-images.githubusercontent.com/51652878/153134978-fc798eb2-71d6-455e-88ec-49d005867ef3.png)

**Steps to reproduce**

1. download ue sources
2. add include path like this:
```
{
    "configurations": [
        {
            "name": "Win32",
            "includePath": [
                "${workspaceFolder}/**",
                "D:/Program Files/Epic Games/UE_4.27/Engine/Source/**"
            ],
            "defines": [
                "_DEBUG",
                "UNICODE",
                "_UNICODE"
            ],
            "windowsSdkVersion": "10.0.19041.0",
            "compilerPath": "C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe",
            "cStandard": "c17",
            "cppStandard": "c++17",
            "intelliSenseMode": "windows-msvc-x64"
        }
    ],
    "version": 4
}
```
3.  Find the file `\Epic Games\UE_4.27\Engine\Source\Runtime\Engine\Private\PlayerController.cpp`
4. Unable to find correct header file location
- `\Epic Games\UE_4.27\Engine\Source\Runtime\Engine\Private\PlayerController.cpp`
![image](https://user-images.githubusercontent.com/51652878/153135588-dd630012-6a63-4358-8606-f5735cf97b23.png)
- `\Epic Games\UE_4.27\Engine\Source\Runtime\Engine\Classes\GameFramework\PlayerController.h`
![image](https://user-images.githubusercontent.com/51652878/153135686-3bb07a4f-5b3e-4ffa-9ee6-34b8375e296a.png)

**Expected behavior**

Can correctly find the header file path

**Comment by Colengms** [comment]
Hi @inaryart .  There is additional configuration (defines, include paths) required to configure IntelliSense for Unreal Engine projects.  In the Unreal Editor, there should be an option to export your project as a VS Code project.  I believe that should generate a `c_cpp_properties.json` and other files (possibly a `compile_commands.json`), which should allow Unreal Engine projects to work with VS Code and C/C++ Extension 'out of the box'.  Could you try that approach and see if it works for you?

**Comment by inaryart** [comment]
It can export the visual studio project, but not the vscode configuration. Anyway, if I use the `"directory/**"` language service should need to automatically find the corresponding path, right? Because his directory does not have any special,

i found an extension:
https://github.com/boocs/ue4-intellisense-fixes

**Comment by Colengms** [comment]
Hi @inaryart .  You will need to go into the settings for Unreal Editor, under General -> Source Code, and select Visual Studio Code as your editor.  That will change the option on the File menu into `Generate Visual Studio Code Project`.

**Comment by inaryart** [comment]
I have exported from ue as a vscode project, still can't find source files.  (I don't know if it's still a language service issue)
- C/C++ configuration  warning:
```
[2022/2/11 10:00:33 AM] Failed to resolve configuration using compilerPath 'C:\Program'. Please use "cl.exe" instead.
[2022/2/11 10:00:33 AM] Failed to resolve configuration using compilerPath 'C:\Program'. Please use "cl.exe" instead.
[2022/2/11 10:00:33 AM] Failed to resolve configuration using compilerPath 'C:\Program'. Please use "cl.exe" instead.
[2022/2/11 10:00:33 AM] Failed to resolve configuration using compilerPath 'C:\Program'. Please use "cl.exe" instead.
[2022/2/11 10:00:34 AM] Failed to resolve configuration using compilerPath 'C:\Program'. Please use "cl.exe" instead.
[2022/2/11 10:00:34 AM] Failed to resolve configuration using compilerPath 'C:\Program'. Please use "cl.exe" instead.
[2022/2/11 10:00:34 AM] Failed to resolve configuration using compilerPath 'C:\Program'. Please use "cl.exe" instead.
[2022/2/11 10:00:34 AM] Failed to resolve configuration using compilerPath 'C:\Program'. Please use "cl.exe" instead.
[2022/2/11 10:00:34 AM] Failed to resolve configuration using compilerPath 'C:\Program'. Please use "cl.exe instead
```

- `.vscode\c_cpp_properties.json`
```
{
    "configurations": [
        {
            "name": "ShooterGameEditor Editor Win64 Development (射击游戏)",
            "intelliSenseMode": "msvc-x64",
            "compileCommands": "D:\\Unreal Projects\\ShooterGame\\.vscode\\compileCommands_射击游戏.json",
            "cStandard": "c17",
            "cppStandard": "c++17"
        },
        {
            "name": "Win32",
            "intelliSenseMode": "msvc-x64",
            "compileCommands": "D:\\Unreal Projects\\ShooterGame\\.vscode\\compileCommands_Default.json",
            "cStandard": "c17",
            "cppStandard": "c++17"
        }
    ],
    "version": 4
}
```
- `.vscode\compileCommands_Default.json`
```
[
	{
		"file": "D:\\\\Unreal Projects\\\\ShooterGame\\\\Source\\\\ShooterGame\\\\Private\\\\Bots\\\\BTDecorator_HasLoSTo.cpp",
		"command": "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Enterprise\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX64\\x64\\cl.exe @\"D:\\Unreal Projects\\ShooterGame\\.vscode\\compileCommands_Default\\ShooterGame.221.rsp\"",
		"directory": "D:\\Program Files\\Epic Games\\UE_4.27\\Engine\\Source"
	},
	{
		"file": "D:\\\\Unreal Projects\\\\ShooterGame\\\\Source\\\\ShooterGame\\\\Private\\\\Bots\\\\BTTask_FindPickup.cpp",
		"command": "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Enterprise\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX64\\x64\\cl.exe @\"D:\\Unreal Projects\\ShooterGame\\.vscode\\compileCommands_Default\\ShooterGame.221.rsp\"",
		"directory": "D:\\Program Files\\Epic Games\\UE_4.27\\Engine\\Source"
	},
...
...
...
```

**Comment by sean-mcmanus** [comment]
@inaryart You appear to be hitting the issue Colen described at https://github.com/microsoft/vscode-cpptools/issues/8812#issuecomment-1035594497 -- are you able to follow his suggestion to fix it?

**Comment by inaryart** [comment]
I tried adding `\"` to the cl path, then changed `C:/Program` to `C:/TEST` (for testing), and found that the warning still contains `C:\Program`, cpptools will generate a configuration cache ? where and can i delete it.

```
  {
    "file": "D:\\\\Unreal Projects\\\\ShooterGame\\\\Source\\\\ShooterGame\\\\Private\\\\Bots\\\\BTDecorator_HasLoSTo.cpp",
    "command": "\"C:/TEST Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.29.30133/bin/HostX64/x64/cl.exe\" @\"D:\\Unreal Projects\\ShooterGame\\.vscode\\compileCommands_Default\\ShooterGame.221.rsp\"",
    "directory": "D:\\Program Files\\Epic Games\\UE_4.27\\Engine\\Source"
  },
```
-  warning
```
[2022/2/11 10:50:38 AM] Failed to resolve configuration using compilerPath 'C:\Program'. Please use "cl.exe" instead.
[2022/2/11 10:50:38 AM] Failed to resolve configuration using compilerPath 'C:\Program'. Please use "cl.exe" instead.
[2022/2/11 10:50:38 AM] Failed to resolve configuration using compilerPath 'C:\Program'. Please use "cl.exe" instead.
```

**Comment by sean-mcmanus** [comment]
It sounds like you may have changed the wrong compile commands file? We read the file directly and there isn't any additional cache or copy of the file. Your c_cpp_properties.json has 2 configurations -- is the correct one active?

**Comment by inaryart** [comment]
> It sounds like you may have changed the wrong compile commands file? We read the file directly and there isn't any additional cache or copy of the file. Your c_cpp_properties.json has 2 configurations -- is the correct one active?

I deleted the `.vscode` folder and then used ue to export the vscode project, but this error still occurs, how can I see more detailed warnings, such as, which file does the warning come from?

**Comment by Colengms** [comment]
Hi @inaryart .  The c_cpp_properties.json` content you posted referred to two compile-commands files:

>"compileCommands": "D:\\Unreal Projects\\ShooterGame\\.vscode\\compileCommands_射击游戏.json",
>"compileCommands": "D:\\Unreal Projects\\ShooterGame\\.vscode\\compileCommands_Default.json",

I think Sean was suggesting that you ensure you make the correction to both of those files.

The correction would be to add (escaped) quotes around the path to `cl.exe`, like so:

Before:
```
"command": "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Enterprise\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX64\\x64\\cl.exe @\"D:\\Unreal Projects\\ShooterGame\\.vscode\\compileCommands_Default\\ShooterGame.221.rsp\"",
```
After:
```
"command": "\"C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Enterprise\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX64\\x64\\cl.exe\" @\"D:\\Unreal Projects\\ShooterGame\\.vscode\\compileCommands_Default\\ShooterGame.221.rsp\"",
```

The warnings that are logged to the "Configuration Warnings" output channel will not automatically be cleared.  If you reload the window after correcting all compile commands file, they should not reoccur.

**Comment by Colengms** [comment]
@inaryart .   The generated project is "multi-root" workspace.  Be sure to apply the same fix to compile commands in the "UE4" workspace folder. 

![image](https://user-images.githubusercontent.com/49173979/153537143-948802d3-ec2e-4d48-84dc-65ea7a5dfece.png)

**Comment by inaryart** [comment]
after my continuous testing, only by manually adding the `compilerPath` option in `.vscode\c_cpp_properties.json`, there will be no warning, but the header file is still not found. Can you find it in your project?
```
{
    "configurations": [
        {
            "name": "ShooterGameEditor Editor Win64 Development (射击游戏)",
            "intelliSenseMode": "msvc-x64",
            "compilerPath": "C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe",
            "compileCommands": "D:\\Unreal Projects\\ShooterGame\\.vscode\\compileCommands_射击游戏.json",
            "cStandard": "c17",
            "cppStandard": "c++17"
        },
        {
            "name": "Win32",
            "intelliSenseMode": "msvc-x64",
            "compilerPath": "C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe",
            "compileCommands": "D:\\Unreal Projects\\ShooterGame\\.vscode\\compileCommands_Default.json",
            "cStandard": "c17",
            "cppStandard": "c++17"
        }
    ],
    "version": 4
}
```
![Code_08MwC0yWMN](https://user-images.githubusercontent.com/51652878/153546045-2f912e18-f84c-4b84-949a-50ce08f16d02.gif)

**Comment by Colengms** [comment]
@inaryart .  I'm able to repro your issue with headers not being found.  This looks to be a problem in the generated project, so a bug in the Unread Editor.  For example, there is no entry for "PlayerController.cpp" in any compile commands file.  As a result, no configuration is being provided for that file when opened.  I see the following configuration warning when opening it:

```
[2/11/2022, 11:16:22 AM] "PlayerController.cpp" not found in "${workspaceFolder}\.vscode\compileCommands_Default.json". 'includePath' from c_cpp_properties.json in folder 'UE_4.27' will be used for this file instead.
```

I'm also seeing other configuration issues.  For example, the file `ShoorterPlayerController.cpp` is unable to locate `ShooterGame.h`.  However, the configuration for that file does not refer to the path (or any parent path) that contains `ShooterGame.h`.

I will try to follow up with Epic Games directly.

**Comment by Colengms** [comment]
I believe the issues here have either been resolved or are external.  It looks like recent versions of Unreal Engine are quoting the compiler path, so spaces in a compiler path should no longer be an issue.  I'm seeing Unreal Engine successfully configure IntelliSense for project source files.  However, I see that Unreal Engine is currently omitting the necessary configuration information (entries in compile_commands.json) for any sources within the core engine itself (such as `PlayerController.cpp`).  Manually adding those entries to the compile_commands.json files (and leveraging the RSP files for each that they provide) appears to result in correct IntelliSense configurations for those files.  I've emailed a contact at Epic about the lack of configuration entries for core engine source files, which may be intentional for some reason.  Addressing that issue would seem to be up to Epic.

Closing this issue.
